### PR TITLE
Expanding providers

### PR DESCRIPTION
**Pull request by yuri**

### CODE REVIEW

**Comment by yuri** [comment]
@SethDavenport Please review retroactively (on Thursday or Friday)

@n1cholasv: FIY

**Comment by yuri** [comment]
@SethDavenport Actually, if you find time to review tomorrow, that would be better.

File: handout/15-providers.md [hunk 1]
```diff
@@ -1,42 +1,92 @@
...
     var service = {};
 
     // Implement the router service as before.
 
     return service;
-  }
+  };
+```
+
+Or, if we prefer, we can use annotations:
```
Comment (SethDavenport) [inline handout/15-providers.md]: Have we actually talked about annotations anywhere?

File: handout/15-providers.md [hunk 2]
```diff
@@ -80,23 +133,134 @@ section's module:
...
+This is not a big problem, the function we are providing won't be called at
+this point either. Rather, it will only be called much later, when service
+router has already been created. This does mean, however, that we'll have to
+provide a reference to the service as an argument to our method.
+
+In other words, when we call `router.tasks.getTaskId()`, we would actually
+have to do this: `router.tasks.getTaskId(router)`. This way, our
+implementation of this method, which we provided in the config block will have
+access to the instance of `router` that we didn't have when we defined (but
+which we do have now). This is uggly, though.
```
Comment (SethDavenport) [inline handout/15-providers.md]: s/uggly/ugly/

File: handout/15-providers.md [hunk 3]
```diff
@@ -80,23 +133,134 @@ section's module:
...
+
+We can fix this by making our implementation of the `$get()` method a bit more
```
Comment (SethDavenport) [inline handout/15-providers.md]: s/more smart/smarter/

File: handout/15-providers.md [hunk 4]
```diff
@@ -80,23 +133,134 @@ section's module:
...
+smart. Let's take advantage of the fact that while router service is not
+available in the scope of the config block, it _is_ available inside `$get()`.
+So, instead of simply copying all methods from `sectionMethods`, let's tweak
+each one of them, inserting passing the service as the first argument. We can
```
Comment (SethDavenport) [inline handout/15-providers.md]: inserting passing - only need one of these words here

**Comment by SethDavenport** [comment]
Some typos, but otherwise :+1:

**Pull request merged**